### PR TITLE
Allow Title Bar to be Hidden (closes #13249)

### LINKED ISSUE #13249: Custom title bar on Windows/Linux
Using dark theme and having white title bar is not nice. Please make it more consistent with rest of the app like in Adobe PS or  Brackets (also look #4599).

### PR DESCRIPTION
**Pull request by TheDrawingCoder-Gamer**
### Requirements for Adding, Changing, or Removing a Feature

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must contribute a change that has been endorsed by the maintainer team. See details in the template below.
* The pull request must update the test suite to exercise the updated functionality. For guidance, please see <https://flight-manual.atom.io/hacking-atom/sections/writing-specs/>.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see <https://github.com/atom/.github/tree/master/CONTRIBUTING.md#pull-requests>.

### Issue or RFC Endorsed by Atom's Maintainers

See https://github.com/atom/atom/issues/13249

### Description of the Change

This opens the already available option on macOS to hide the title bar to Windows and Linux. This requires little work and makes it much easier for packages like atom-title-bar-replacer to work. Note only the hidden and native options are exposed.

### Alternate Designs

Editing the source code like atom-title-bar-replacer does. This is undesirable as source editing is dangerous and unstable

### Possible Drawbacks

 The title bar being hidden might make it hard to close the window. 

### Verification Process

Toggling the title bar on and off seems to cause no issues. 

### Release Notes
Allows Hiding of Title Bar on all platforms

### CODE REVIEW

**Comment by TheDrawingCoder-Gamer** [comment]
It seems the linter isn't working. Any help would be appreciated!

**Comment by TheDrawingCoder-Gamer** [comment]
This closes #13249

**Comment by DanielTamkin** [comment]
Oh this is so exciting @TheDrawingCoder-Gamer!

**Comment by TheDrawingCoder-Gamer** [comment]
It seems that the specs aren't working. Please help with specs so we can hammer this out!

**Comment by TheDrawingCoder-Gamer** [comment]
Please help me unbork this

**Comment by DanielTamkin** [comment]
I'm pretty limited in my knowledge about the backend of Atom, but i'm sure i could find some time soon to learn a bit of it. Might be able to to give you a hand @TheDrawingCoder-Gamer. Is the linter failing and that's it?

**Comment by TheDrawingCoder-Gamer** [comment]
No, some more is borked. Some things in the text editor in spec are failing. Run specs, mostly specs are borked

**Comment by DeeDeeG** [comment]
Hi @TheDrawingCoder-Gamer, fellow Atom hackerizer from the community here.

I'd like to point out that some tests fail by timing out in the CI environment. They often fail even for "correct" code, and pass when allowed to re-run several times. Are you certain that the specs are truly broken? Maybe they just need a chance to re-run.

Also, NSlog builds okay for me on `master` branch. It does have native code in it, so you have to be somewhat careful to build it cleanly. You can make sure everything is built cleanly by doing `script/clean` then `script/bootstrap` and finally `script/build --no-bootstrap`, though that does take a while.

**Comment by DeeDeeG** [comment]
There are a fair amount of changes here that don't look 100% related to this PR. You can revert them one by one by doing the following:

(For any version of git): `git checkout c79d2adbe0c8902e87624049982a31fd7e4c3ebe -- path/to/file`

(For git 2.25 or newer): `git restore --source="c79d2adbe0c8902e87624049982a31fd7e4c3ebe" -- path/to/file`. For example: `git restore --source="c79d2adbe0c8902e87624049982a31fd7e4c3ebe" -- apm/package-lock.json`

**Comment by TheDrawingCoder-Gamer** [comment]
I'm using GitHub Desktop and revert commit seems to be working well. I wish I had made more commits or I wouldn't have to be unstaging hunks.

**Comment by TheDrawingCoder-Gamer** [comment]
I'm sure some tests are messed up because there are number mismatches, and the like.

**Comment by TheDrawingCoder-Gamer** [comment]
I think this would actually pass but there is a timeout

**Comment by TheDrawingCoder-Gamer** [comment]
Can I force the check to rerun or does a mod have to

**Comment by DeeDeeG** [comment]
I see there are a couple of timeouts on Windows x64, which isn't really that uncommon at this repo. CI timeouts on genuinely working code tend to happen on Windows x64, or one of the three macOS test runners. (It's a problem that's being looked at and worked on, but the cause isn't obvious, so the solution isn't obvious either.)

https://github.visualstudio.com/Atom/_build/results?buildId=80338&view=logs&j=4fbced83-508e-5fe0-c978-5c71ec0fc506&t=b7851fd1-a7b6-5bbd-e24e-1d32b43df7d7&l=2433

So yeah, it's possible this would pass. You can set up a (free) Azure DevOps account if you want to run the CI. (I already have an account set up, so I could run it for you if that would help, to see if it passes.)

> Can I force the check to rerun or does a mod have to

I'm pretty sure one of Atom's maintainers would have to do, for the CI attached to this PR. Anyone can clone your branch and run CI on it from their own Azure DevOps account, though. (That takes a bit of setup time, I'll admit).

**Comment by DeeDeeG** [comment]
Also, you can add a small commit that doesn't do much if you just want to trigger CI. If you do `git commit --allow-empty` from the command-line, that will produce a no-changes commit. Pushing that updated commit would trigger CI to run again.

It's a bit less efficient than a maintainer re-running the build, becase maintainers can click "re-run failed jobs" to only rerun the failing tests. Takes less time, and doesn't risk one of the macOS jobs randomly failing.

**Comment by TheDrawingCoder-Gamer** [comment]
Something tells me I set [this](https://dev.azure.com/benharless820/benharless820/_build/results?buildId=5&view=logs&j=e2cf4b02-5697-54ad-cf7c-fc2a840d53af) up improperly

**Comment by DeeDeeG** [comment]
@TheDrawingCoder-Gamer oh yeah, I forgot  that still needs to be manually set up to make things work. I and another person have been trying to make it mostly work "out of the box" but there is a little piece that still needs to be manually done.

Go to here: https://github.com/settings/tokens
- Press the "Generate new token" button
- Leave all boxes unchecked, you just want the token to be able to access the API, not post to your repos!
- Enter a name for the token and press the "Generate token" button at the bottom of the page
- Make sure you grab the token before navigating away from this page, or you'll have to regenerate it (which to be fair only takes a second).

Open Azure DevOps (perhaps in another browser tab so you don't lose the token).
- Visit the "Pipelines" view (easy ways to get there are from the sidebar, or from the "breadcrumb trail" at the top of the screen)
- Click the pipeline you want to edit
- Press the Edit button.
- Press the "Variables" button
- Add the token from before as a secret variable `GITHUB_TOKEN` (all caps).
- You may want to repeat adding the token to your other pipelines, if you have more than one pipeline going.

**Comment by TheDrawingCoder-Gamer** [comment]
https://dev.azure.com/benharless820/benharless820/_build/results?buildId=7&view=logs&j=50a69d0a-7972-5459-cdae-135ee6ebe312&t=50a69d0a-7972-5459-cdae-135ee6ebe312 Epic

**Comment by TheDrawingCoder-Gamer** [comment]
Do I have to fix the CI build to get a pull request merged.

**Comment by TheDrawingCoder-Gamer** [comment]
Should I try to reroll the dice or is this okay as is?

**Comment by TheDrawingCoder-Gamer** [comment]
https://github.com/atom/atom/blob/master/CONTRIBUTING.md#pull-requests This says don't reroll.

**Comment by TheDrawingCoder-Gamer** [comment]
Who would be good to @?

**Comment by DeeDeeG** [comment]
I think @sadick254 and @darangi are the most active at the moment?

To those I just `@mentioned`: Hi folks, this is a pull request I have been helping out with. Please take a look if you have some time. Thanks.

**Comment by sadick254** [comment]
@TheDrawingCoder-Gamer Just to let you know. I have been following this PR closely. It seems to contain a lot of unrelated changes.

**Review by sadick254 — CHANGES_REQUESTED** [review]
I think this is ready only that there are a lot of unrelated changes.
### Next steps
- [ ] Revert changes made to `apm/package-lock.json`
- [ ] Revert change made to `package-lock.json`
- [ ] Revert change made to `script/package-lock.json`
- [ ] Revert changes made to `spec/helpers/words.js`

**Comment by TheDrawingCoder-Gamer** [comment]
Reverting words.js to commit c79d2adbe0c8902e87624049982a31fd7e4c3ebe doesn't seem to be working

**Comment by TheDrawingCoder-Gamer** [comment]
Does this mean it's equivalent to the commit?

**Comment by DeeDeeG** [comment]
@TheDrawingCoder-Gamer If I had to guess, the difference of `words.js` between this PR and before this PR... is probably line-endings. `CRLF` is the line-ending on Windows. `LF` is the line ending on Linux and macOS.

Line endings aren't always easy to notice. Even if you can't see the change when restoring that file, are you able to stage and commit it after restoring it?

**Comment by TheDrawingCoder-Gamer** [comment]
I'm pretty sure github desktop auto converts line endings so it doesn't matter anyways.

**Comment by TheDrawingCoder-Gamer** [comment]
I think I borked words.js when I ran run prettier. I reverted run prettier then a conflict emerged, which I resolved. It must think there are still changes?

**Comment by TheDrawingCoder-Gamer** [comment]
When I restore words.js it doesn't pop up in GitHub Desktop

**Comment by DeeDeeG** [comment]
In terms of how git saves files, the different line endings mean saving an entirely separate copy of a file that's 46,891 lines long, and ~643kB on disk.

So, if it's possible to revert the change, that would still be worthwhile.

Maybe this is an uncommon situation where GitHub Desktop can't help?

On command-line I would do:

```
git restore --source="c79d2adbe0c8902e87624049982a31fd7e4c3ebe" -- spec/helpers/words.js
git add spec/helpers/words.js
git commit -m "[some commit message here]"
```

**Comment by TheDrawingCoder-Gamer** [comment]
```
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
```

**Comment by DeeDeeG** [comment]
Does this help?

Windows:

```
del spec/helpers/words.js
git checkout c79d2adbe0c8902e87624049982a31fd7e4c3ebe -- spec/helpers/words.js
```
Linux or macOS: 

```
rm spec/helpers/words.js
git checkout c79d2adbe0c8902e87624049982a31fd7e4c3ebe -- spec/helpers/words.js
```

**Comment by TheDrawingCoder-Gamer** [comment]
Working tree clean

**Comment by DeeDeeG** [comment]
If that didn't work, then you can open the file in Atom. Then at the bottom are a row of buttons `[CR]LF, UTF-8, Plain Text, [branch-name], Fetch, GitHub, Git ([number])`

You can click "LF" (or "CRLF") and switch to the opposite line ending. (If it's LF, switch to CRLF. If it's CRLF, switch to LF). Then save the file. Then you should be able to use the "Git" tab of Atom to stage the change. And also to commit the change.

**Comment by TheDrawingCoder-Gamer** [comment]
Is this all necessary? I think it's clear that there are no differences.

**Comment by DeeDeeG** [comment]
I leave it to Atom's maintainers.

Technically the PR re-writes the entire `words.js` file, which adds a little bit of noise to the commit history of Atom. But either line ending should work. Node can handle either line ending when reading/executing JS files, so the practical impact should be zero.

**Comment by TheDrawingCoder-Gamer** [comment]
How about I add a commit where I change 1 letter in `words.js` then another where I restore to commit c79d2adbe0c8902e87624049982a31fd7e4c3ebe

**Comment by DeeDeeG** [comment]
I suppose that's fine, as long as it works. (I shouldn't really answer that, since it's the Atom maintainers' repo, so it would be down to their preferences. But looking at past PRs, things like that have been merged.)

**Comment by TheDrawingCoder-Gamer** [comment]
Running `git diff c79d2adbe0c8902e87624049982a31fd7e4c3ebe --ignore-cr-at-eol -- .` doesn't include words.js so

**Comment by DeeDeeG** [comment]
I tried this from Windows and nothing I tried worked. I was able to fix it from a Linux computer, though.

https://github.com/atom/atom/commit/41eb5a5e874b6ab60b6ede0c4f813a94991dced1

I can open a pull request to your branch over at your fork of Atom if you'd like?

**Comment by TheDrawingCoder-Gamer** [comment]
Yes please I am getting frustrated over this file

**Comment by DeeDeeG** [comment]
It looks like the changes to that file are resolved now.

@TheDrawingCoder-Gamer Sorry about all the frustration.

Usually I try to help people learn the git commands (that old saying about "teach the person to fish rather than give them the fish"). But in this case I don't know if this was possible to fix on Windows at all. So there was no point trying to walk you or anyone through the process. If I had known that I would have offered to open a PR sooner.

**Comment by TheDrawingCoder-Gamer** [comment]
It is more helpful to learn than be given, but I never knew that this could be unfixable. We're in the home stretch now, and it's probably gonna be included. Thank you for helping with this PR!

File: spec/git-repository-spec.js
Comment (sadick254) [inline spec/git-repository-spec.js:407]: Is this change necessary?

Comment (TheDrawingCoder-Gamer) [reply spec/git-repository-spec.js:407]: Yes, without this a spec fails because someone renamed a file without editing this.

Comment (Aerijo) [reply spec/git-repository-spec.js:407]: That doesn't sound right. Surely the tests would be consistently failing for this reason on master if that were the case. `atom.workspace.open` should work even on files that don't exist, and there are several other references to `file.txt` anyway.

Comment (TheDrawingCoder-Gamer) [reply spec/git-repository-spec.js:407]: I need to take a closer look. It is in `working-dir` correct?

Comment (TheDrawingCoder-Gamer) [reply spec/git-repository-spec.js:407]: a.txt seems to have no content...

Comment (DeeDeeG) [reply spec/git-repository-spec.js:407]: Maybe this test failed just due to flakiness? Other branches and PRs do pass with this saying `file.txt`.

Comment (TheDrawingCoder-Gamer) [reply spec/git-repository-spec.js:407]: Ah. Will undo this.

Comment (Aerijo) [reply spec/git-repository-spec.js:407]: That sounds most likely. Is there a link to the test results that did fail with it?

Comment (TheDrawingCoder-Gamer) [reply spec/git-repository-spec.js]: No. I'm running a clone of the pipeline on my version right now though.

Comment (TheDrawingCoder-Gamer) [reply spec/git-repository-spec.js]: Resolving because this is outdated...

File: spec/text-editor-component-spec.js
Comment (TheDrawingCoder-Gamer) [inline spec/text-editor-component-spec.js:3775]: All this file changes is formatting. Discarding this would make the Run Linter fail, right?

Comment (TheDrawingCoder-Gamer) [reply spec/text-editor-component-spec.js:3775]: https://prettier.io/docs/en/rationale.html#multi-line-objects This doesn't make sense

Comment (Aerijo) [reply spec/text-editor-component-spec.js:3775]: Its saying it preserves the single/multi line layout to a reasonable extent. So you can revert the change and the lint should still pass. If in doubt, just run `script/lint`.

Comment (TheDrawingCoder-Gamer) [reply spec/text-editor-component-spec.js:3775]: Okay.

Comment (DeeDeeG) [reply spec/text-editor-component-spec.js:3775]: By the way, in my experience, CI doesn't error out on any "badly formatted code" in files you didn't change in your PR.

(I don't know how this works exactly; Maybe some persistent cache of files that were already linted once and deemed okay? Or it's clever enough to look at only changed files from reading the git metadata?)

**Comment by TheDrawingCoder-Gamer** [comment]
https://dev.azure.com/benharless820/benharless820/_build/results?buildId=9&view=results Restore git-repository-spec.js

**Comment by TheDrawingCoder-Gamer** [comment]
Linux failed 3 times

**Comment by sadick254** [comment]
@TheDrawingCoder-Gamer The linux failures are unrelated to this PR and we are trying to figure out why its failing https://github.com/atom/apm/pull/900. Thank you for your contribution. Looking forward to review more PR's from you :smile:

**Review by sadick254 — APPROVED** [review]

**Comment by TheDrawingCoder-Gamer** [comment]
https://dev.azure.com/benharless820/benharless820/_build/results?buildId=10&view=results Success!

**Comment by tonglei100** [comment]
How to hidden the title bar? I'm use the 1.53.0 on Windows 10

**Comment by ThatXliner** [comment]
> How to hidden the title bar? I'm use the 1.53.0 on Windows 10

On a Mac, you can go to core > title bar in the settings.

**Comment by tonglei100** [comment]
> > How to hidden the title bar? I'm use the 1.53.0 on Windows 10
> 
> On a Mac, you can go to core > title bar in the settings and edit it from there

Thank you! on Windows 10 it's OK too

**Comment by DeeDeeG** [comment]
@tonglei100 

>But when I hidden the title bar and menu, I can't show the menu. Now how can I do to use the menu?

When this setting is enabled, Atom's built-in menu goes away. You can still visit settings by pressing <kbd>Ctrl</kbd> + <kbd>,</kbd>.

This is meant to be used with something that provides a custom title bar, like this: https://atom.io/packages/title-bar-replacer (I'm sure there are other packages like available as well.)

**Comment by tonglei100** [comment]
> @tonglei100
> 
> > But when I hidden the title bar and menu, I can't show the menu. Now how can I do to use the menu?
> 
> When this setting is enabled, Atom's built-in menu goes away. You can still visit settings by pressing Ctrl + ,.
> 
> This is meant to be used with something that provides a custom title bar, like this: https://atom.io/packages/title-bar-replacer (I'm sure there are other packages like available as well.)

title-bar-replacer is very good! thank u

**Pull request merged**
